Support Cross encoder models

maxdebayser opened the pull request:
This PR adds support for Cross Encoder models reusing most of what was done to support embedding models. It includes:
- Support for RoBERTA SequenceClassification models, tested with `BAAI/bge-reranker-v2-m3`
- Support for BERT SequenceClassification models, tested with `cross-encoder/ms-marco-MiniLM-L-6-v2`
- A `score` API endpoint
- A `LLM.score()` method
- Support for GPU and CPU

An example of an API call is:

```bash
$ curl -X 'POST'   'http://localhost:8000/v1/score'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{
  "model": "BAAI/bge-reranker-v2-m3",
  "text_1": "What is the capital of France?",
  "text_2" : ["The capital of France is Paris."]
}
```

Response:
```json
{
  "id": "score-64dcf95966094892b0f0ccc63637c6d2",
  "object": "list",
  "created": 239128,
  "model": "BAAI/bge-reranker-v2-m3",
  "data": [
    {
      "index": 0,
      "object": "score",
      "score": [
        1.0
      ]
    }
  ],
  "usage": {
    "prompt_tokens": 0,
    "total_tokens": 0,
    "completion_tokens": 0,
    "prompt_tokens_details": null
  }
}

```

Instead of adding a new task type, the models in this PR are loaded with `task == "embedding"` but support the additional `is_cross_encoder` property. This allows us to reuse most of what was done the embedding support. The only changes required to the core layers between the `API` and `LLM` is the propagation the `token_type_ids` list. This is an additional tokenizer output that is generated for the BERT models and has to be passed to the model's embedding layer.

cc: @DarkLight1337 @flaviabeo 

FIX #8022

maxdebayser committed:
Add support for cross encoders

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser committed:
Merge branch 'main' into cross_encoder

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

DarkLight1337 commented on vllm/model_executor/models/bert.py:
```diff
@@ -426,3 +451,96 @@ def _build_pooler(self, pooler_config: PoolerConfig) -> Pooler:
+            for name, weight in weights:
+                if name.startswith("bert."):
+                    yield (name[5:], weight)
+                    #bert_weights.append((name[5:], weight))
```
```suggestion
```

DarkLight1337 replied:
Did you forget to remove this?

DarkLight1337 reviewed (commented):
Can you add an example script to show how to use this API?

QQ: Why do we have to define a separate `"cross_encoding"` task for this? I think we can keep using `"embedding"` task if we just override the `pooler` method instead of defining a new `classification_output` method.

maxdebayser commented:
> Can you add an example script to show how to use this API?

Yes, I've added one in the PR description but it's a good idea to add it to the documentation.

> QQ: Why do we have to define a separate `"cross_encoding"` task for this? I think we can keep using `"embedding"` task if we just override the `pooler` method instead of defining a new `classification_output` method.

I thought about this, and the only reason I kept it that way was that in the serving layer I need to know what task is being done because I need to call the tokenizer with `tokenizer(text=text1, text_pair=text2)`. If instead of reusing the chat embeddings API we had a new endpoint just for cross encoding, this wouldn't be necessary. Or perhaps we can add an attribute to some of the config classes to tell the although the task is "embedding", the model is actually a "BertModelForSequenceClassification"

DarkLight1337 commented:
I think it may be better to make this a separate flag, similar to how we have a flag for multimodal models, rather than creating a new task for it.

DarkLight1337 commented:
I think it may be better to make this a separate flag, similar to how we have a flag for multimodal models, rather than creating a new task for it.

DarkLight1337 commented:
I think having a separate API for this would be cleaner as well - maybe a Scoring API where we output a single score?

DarkLight1337 commented:
I think having a separate API for this would be cleaner as well - maybe a Scoring API where we output a single score?

maxdebayser committed:
Add support for roberta models, including BAAI/bge-reranker-v2-m3

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser committed:
remove task cross_encoding

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser commented:
I've removed the "cross_encoding" task and added a `is_cross_encoder` property to the `ModelConfig` class. I've also added support for Roberta models.

Pending TODOs:
- Add Scoring API
- Support for cross encoding in the LLM class
- Tests comparing with sentence-transformers
- CPU support

DarkLight1337 commented on vllm/model_executor/models/interfaces.py:
```diff
@@ -350,3 +351,57 @@ def is_attention_free(
+
+
+@runtime_checkable
+class SupportsCrossEncoding(Protocol):
```
Let's use a classvar here as a tag.

maxdebayser committed:
address review comments, fix bug, clean up diff

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser committed:
add cpu support

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser committed:
Add a score() method top the LLM entrypoint

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser committed:
raise exception in case of MistralTokenizer

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser committed:
Add tests for the LLM.score() method

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser commented:
I've added a `score()` method to the LLM class and added tests for it. I've also fixed the CPU support.

Pending TODOs:

- Add Scoring API
- Test Scoring API comparing with sentence-transformers

DarkLight1337 commented on vllm/entrypoints/llm.py:
```diff
@@ -806,24 +806,26 @@ def encode(
         self,
-        query: SingletonPrompt,
         texts: Union[SingletonPrompt, Sequence[SingletonPrompt]],
+        text_pairs: Union[SingletonPrompt, Sequence[SingletonPrompt]],
```
Maybe `text_1` and `text_2`? `text_pairs` is a bit confusing to me as it suggests that I should pass in a list of tuples.

maxdebayser committed:
refactor common code

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

DarkLight1337 commented on vllm/transformers_utils/config.py:
```diff
@@ -568,3 +570,12 @@ def try_get_generation_config(
+def get_cross_encoder_activation_function(config: PretrainedConfig):
+    if (hasattr(config, "sbert_ce_default_activation_function")
+            and config.sbert_ce_default_activation_function is not None):
+        return import_from_string(
+            config.sbert_ce_default_activation_function)()
```
For better security, perhaps we should limit the modules that can be imported this way? Though I guess it may break compatibility with sentence-transformers...

maxdebayser committed:
do the AMD lazy loading thing

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser committed:
load activation functions only from pytorch

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser commented:
This PR already supports the `BAAI/bge-reranker-v2-m3` model https://github.com/vllm-project/vllm/issues/8022

maxdebayser committed:
positional-only arguments

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

flaviabeo committed:
Adds /v1/score route

- Texts lists input protocol + score responses endpoint
- Supports 1-1, 1-N, N-N requests

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

maxdebayser committed:
Merge branch 'upstream_main' into cross_encoder

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

DarkLight1337 commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -136,31 +138,34 @@ async def create_embedding(
-                    request.input,
-                    truncate_prompt_tokens=truncate_prompt_tokens,
-                    add_special_tokens=request.add_special_tokens,
-                )
+                if isinstance(request, EmbeddingChatRequest):
+                    (
+                        _,
+                        request_prompts,
+                        engine_prompts,
+                    ) = await self._preprocess_chat(
+                        request,
+                        tokenizer,
+                        request.messages,
+                        chat_template=request.chat_template
+                        or self.chat_template,
+                        chat_template_content_format=self.
+                        chat_template_content_format,
+                        add_generation_prompt=request.add_generation_prompt,
+                        continue_final_message=request.continue_final_message,
+                        truncate_prompt_tokens=truncate_prompt_tokens,
+                        add_special_tokens=request.add_special_tokens,
+                    )
+                else:
+                    request_prompts, engine_prompts = self\
+                        ._preprocess_completion(
+                        request,
+                        tokenizer,
+                        request.input,
+                        truncate_prompt_tokens=truncate_prompt_tokens,
+                        add_special_tokens=request.add_special_tokens,
+                    )
```
You forgot to dedent this.

DarkLight1337 commented on vllm/outputs.py:
```diff
@@ -63,9 +63,9 @@ class EmbeddingOutput:
-    def __repr__(self) -> str:
-        return (f"EmbeddingOutput("
-                f"embedding={len(self.embedding)})")
+    # def __repr__(self) -> str:
+    #     return (f"EmbeddingOutput("
+    #             f"embedding={len(self.embedding)})")
```
Is this intended?

flaviabeo replied:
No! Sorry

DarkLight1337 commented on vllm/entrypoints/openai/protocol.py:
```diff
@@ -808,6 +808,29 @@ def to_pooling_params(self):
 
 
+class ScoreRequest(OpenAIBaseModel):
+    # Ordered by official OpenAI API documentation
+    # https://platform.openai.com/docs/api-reference/embeddings
```
Probably should remove this comment

flaviabeo committed:
Adds score unit tests

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

maxdebayser committed:
fix unused code

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser committed:
fix access to None in error handler

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser committed:
fix test code

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser commented:
@DarkLight1337 , thank you so much for guiding this PR along the way. It's ready for review now

DarkLight1337 commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -565,6 +589,12 @@ def init_app_state(
+        model_config,
+        base_model_paths,
+        request_logger=request_logger
+    ) if model_config.task == "embedding" else None
```
Also check cross-encoder here?

DarkLight1337 commented on vllm/model_executor/models/registry.py:
```diff
@@ -203,6 +215,7 @@ def from_model_cls(model: Type[nn.Module]) -> "_ModelInfo":
         return _ModelInfo(
             is_text_generation_model=is_text_generation_model(model),
             is_embedding_model=is_embedding_model(model),
+            supports_cross_encoding=supports_cross_encoding(model),
```
Can you update model registry tests to check this?

DarkLight1337 reviewed (commented):
Overall the code looks good. Please update the OpenAI serving docs to include this new API!

DarkLight1337 commented:
Also @robertgshaw2-neuralmagic are we ready to add BERT and RoBERTa to the list of supported models in the docs?

DarkLight1337 commented on vllm/utils.py:
```diff
@@ -1562,6 +1562,44 @@ def import_from_path(module_name: str, file_path: Union[str, os.PathLike]):
     return module
 
 
+# Copied from sentence_transformers
+def import_from_string(dotted_path: str) -> type:
+    """
+    Import a dotted module path and return the attribute/class designated by 
+    the last name in the path. Raise ImportError if the import failed.
+
+    Args:
+        dotted_path (str): The dotted module path.
+
+    Returns:
+        Any: The attribute/class designated by the last name in the path.
+
+    Raises:
+        ImportError: If the import failed.
+
+    Example:
+        >>> import_from_string(
+                'sentence_transformers.losses.MultipleNegativesRankingLoss')
+    """
+    try:
+        module_path, class_name = dotted_path.rsplit(".", 1)
+    except ValueError as e:
+        msg = f"{dotted_path} doesn't look like a module path"
+        raise ImportError(msg) from e
+
+    try:
+        module = importlib.import_module(dotted_path)
+    except Exception:
+        module = importlib.import_module(module_path)
+
+    try:
+        return getattr(module, class_name)
+    except AttributeError as e:
+        msg = f'Module "{module_path}" does not define a "{class_name}"'
+        ' attribute/class'
+        raise ImportError(msg) from e
```
Let's instead update and use  `resolve_obj_by_qualname` (added in a recent commit).

maxdebayser committed:
Merge branch 'main' into cross_encoder

maxdebayser committed:
use resolve_obj_by_qualname

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser committed:
verify cross encoder

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser committed:
add registry tests

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

DarkLight1337 commented on tests/models/test_registry.py:
```diff
@@ -56,6 +58,17 @@ def test_registry_is_multimodal(model_arch, is_mm, init_cuda):
+    ("RobertaForSequenceClassification", True),
+    ("XLMRobertaForSequenceClassification", True),
+])
+def test_registry_is_cross_encoder(model_arch, is_ce):
+    assert ModelRegistry.is_cross_encoder_model(model_arch) is is_ce
```
Can we merge this test with the one above?

flaviabeo committed:
Adds API usage example at the docs

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

flaviabeo committed:
Merge test registry + fix example lint

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

flaviabeo committed:
yapf disble - conflicts with isort

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

DarkLight1337 commented on tests/models/test_registry.py:
```diff
@@ -6,6 +6,8 @@
 from vllm.model_executor.models import (is_embedding_model,
                                         is_text_generation_model,
                                         supports_multimodal)
+# yapf conflicts with isort for this block
+# yapf: disable
```
You should re-enable yapf afterwards.

DarkLight1337 reviewed (approved):
Otherwise looks good, thanks for your effort and patience!

DarkLight1337 commented:
Wait, we still need to update the OpenAI page under `/docs`

maxdebayser commented:
The tests that aren't failing do to `hugginface.co` timeout are doing so because of a JSON serialization problem. I've opened a tentative fix here: https://github.com/vllm-project/vllm/pull/10580

flaviabeo committed:
Add Cross Encoders score API docs for OpenAI compatible page

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

flaviabeo commented:
@DarkLight1337 I have included in this last commit of mine, some documentation for the score API in the [OpenAI Compatible Server page](https://github.com/maxdebayser/vllm/commit/024837ba7f972f62b13b55b35c4daee63bf90372). Would this be the right place to add it?

maxdebayser committed:
Merge branch 'upstream_main' into cross_encoder

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

DarkLight1337 commented:
#10581 should fix it, please merge from main again.

maxdebayser committed:
Merge branch 'upstream_main' into cross_encoder

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

DarkLight1337 commented:
> @DarkLight1337 I have included in this last commit of mine, some documentation for the score API in the [OpenAI Compatible Server page](https://github.com/maxdebayser/vllm/commit/024837ba7f972f62b13b55b35c4daee63bf90372). Would this be the right place to add it?

I think we can move this up one section so that parameters are shown after all API mentions.

flaviabeo committed:
Moves score API section upMoves score API section upMoves score API section upMoves score API section upMoves score API section upMoves score API section upMoves score API section upMoves score API section upMoves score API section up

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>
